### ISSUE TITLE
Interpretation of expl3 code

### ISSUE THREAD

**Issue by jpgoelz** [texstudio-org/texstudio#61]
Sometimes, the use of `expl3` code within a document is needed, as many packages are written with it.

When the package `expl3` is loaded, the use of this code is started with `\ExplSyntaxOn` and ended with `\ExplSyntaxOff`.
If possible, it should be interpreted differently, as there are a few symbols that have a different meaning as in "normal" LaTeX. [Details are explained in this (concise) answer to a tex.stackexchange.com question.][1] At least, maybe `_` could be ignored, as it makes for some messy looking code:

![expl3](https://user-images.githubusercontent.com/32487587/37400686-f608466a-2785-11e8-84fa-789718f90922.JPG)

[1]: https://tex.stackexchange.com/a/108703/35936

**Comment by 3isenHeiM** [comment]
Relates TeX.SE question: https://tex.stackexchange.com/q/662123/28926

**Comment by schtandard** [comment]
About two years ago I tackled this for highlight.js and [this](https://github.com/highlightjs/highlight.js/blob/main/src/languages/latex.js#L33-L54) is what we came up with:
```javascript
  const L3_REGEX = new RegExp([
    // A function \module_function_name:signature or \__module_function_name:signature,
    // where both module and function_name need at least two characters and
    // function_name may contain single underscores.
    '(?:__)?[a-zA-Z]{2,}_[a-zA-Z](?:_?[a-zA-Z])+:[a-zA-Z]*',
    // A variable \scope_module_and_name_type or \scope__module_ane_name_type,
    // where scope is one of l, g or c, type needs at least two characters
    // and module_and_name may contain single underscores.
    '[lgc]__?[a-zA-Z](?:_?[a-zA-Z])*_[a-zA-Z]{2,}',
    // A quark \q_the_name or \q__the_name or
    // scan mark \s_the_name or \s__vthe_name,
    // where variable_name needs at least two characters and
    // may contain single underscores.
    '[qs]__?[a-zA-Z](?:_?[a-zA-Z])+',
    // Other LaTeX3 macro names that are not covered by the three rules above.
    'use(?:_i)?:[a-zA-Z]*',
    '(?:else|fi|or):',
    '(?:if|cs|exp):w',
    '(?:hbox|vbox):n',
    '::[a-zA-Z]_unbraced',
    '::[a-zA-Z:]'
  ].map(pattern => pattern + '(?![a-zA-Z:_])').join('|'));
```
When a backslash is found, this large regex (`(?![a-zA-Z:_])` is appended to every list entry and then they are all joined using `|`) is applied that matches everything that follows LaTeX3 naming conventions. If it doesn't match, then non-LaTeX3 variants are tried (double-caret syntax, control words or control symbols with `@` considered a letter). This has been in use on tex.sx since then and seems to be working really well.

It has the advantage of not needing any complicated detection of `\ExplSyntaxOn`/`\ExplSyntaxOff`, just like `\makeatletter`/`\makeatother` is currently not detected in favor of just always treating `@` as a letter.

Maybe the same approach can be used for TeXstudio?

**Comment by muzimuzhi** [comment]
According to https://github.com/texstudio-org/texstudio/issues/3491#issuecomment-1985400474 posted on Mar 8, 2024

> basic highlighting for ltx3 is implemented on 4.7.4 (dev)

So perhaps this issue is more-or-less resolved.
